File: apps/mobile/ios/EdecanKit/Sources/EdecanKit/SpeechTags.swift
import Foundation

/// Speech tags y efectos de Eleven v3. No hay lista blanca: `[thoughtfully]`,
/// `[laughs]`, `[applause]`, `[clears throat]` o lo que invente el modelo
/// viaja intacto hasta el TTS. El chat las oculta todas.
public enum SpeechTags {
    /// `[efecto]` de una línea. No come Markdown `[texto](url)` ni `![alt](url)`.
    private static let tag = try? NSRegularExpression(
        pattern: #"\[(?:(?![^\]]{0,120}\]\()[^\]\n]{1,120})\]"#
    )
    private static let espacios = try? NSRegularExpression(pattern: #"[ \t]{2,}"#)
    private static let corte = try? NSRegularExpression(pattern: #"[.!?;…]+[ \t]+|\n{2,}"#)
    private static let valla = try? NSRegularExpression(
        pattern: #"```[\s\S]*?```|!\[[^\]]*\]\([^)]+\)|\[[^\]]+\]\([^)]+\)|`[^`]+`"#
    )

    private static let primera = "warmly"
    private static let rotacion = ["pause", "thoughtful", "calm", "curious", "reassuring", "gently"]

    /// Texto para pintar, copiar o compartir. El original (con tags) se
    /// manda a `Escuchar`.
    public static func ocultar(_ texto: String) -> String {
        let sinTags = reemplazarTags(en: texto, con: "")
        guard let espacios else { return recortarInicio(sinTags) }
        let rango = NSRange(sinTags.startIndex..., in: sinTags)
        let colapsado = espacios.stringByReplacingMatches(
            in: sinTags, range: rango, withTemplate: " "
        )
        return recortarInicio(colapsado)
    }

    /// Devuelve el texto limpio, sin tags inventadas ni speech tags.
    public static func enriquecer(_ texto: String) -> String {
        ocultar(texto)
    }

    // MARK: - Detección

    private static func primerTag(en texto: String) -> String? {
        guard let tag, let match = tag.firstMatch(
            in: texto, range: NSRange(texto.startIndex..., in: texto)
        ), let rango = Range(match.range, in: texto) else { return nil }
        let crudo = texto[rango]
        guard crudo.count >= 3 else { return nil }
        return String(crudo.dropFirst().dropLast()).lowercased()
    }

    private static func reemplazarTags(en texto: String, con plantilla: String) -> String {
        guard let tag else { return texto }
        let ns = texto as NSString
        var piezas: [String] = []
        var cursor = 0
        tag.enumerateMatches(
            in: texto, range: NSRange(location: 0, length: ns.length)
        ) { match, _, _ in
            guard let match else { return }
            let start = match.range.location
            if start > 0, ns.character(at: start - 1) == 33 { return }
            var from = start
            while from > cursor {
                let prev = from - 1
                let ch = ns.character(at: prev)
                if ch == 32 || ch == 9 { from = prev } else { break }
            }
            if from > cursor {
                piezas.append(ns.substring(with: NSRange(location: cursor, length: from - cursor)))
            }
            if !plantilla.isEmpty { piezas.append(plantilla) }
            cursor = match.range.location + match.range.length
        }
        if cursor < ns.length {
            piezas.append(ns.substring(from: cursor))
        }
        return piezas.joined()
    }

    private static func recortarInicio(_ texto: String) -> String {
        String(texto.drop(while: { $0 == " " || $0 == "\t" }))
    }

    // MARK: - Relleno

    private static func elegirTag(sentencia: String, previa: String?, primera: Bool) -> String {
        let limpia = ocultar(sentencia)
        if limpia.contains("?") || limpia.contains("¿") { return distinta("curious", de: previa) }
        if limpia.contains("!") || limpia.contains("¡") { return distinta("excited", de: previa) }
        if primera { return distinta(Self.primera, de: previa) }
        for tag in rotacion where tag != previa { return tag }
        return "thoughtful"
    }

    private static func distinta(_ tag: String, de previa: String?) -> String {
        if tag != previa { return tag }
        return tag == primera ? "calm" : primera
    }

    private static func tramos(en texto: String) -> [Range<String.Index>] {
        guard let corte, !texto.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else {
            return texto.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty ? [] : [texto.startIndex..<texto.endIndex]
        }
        var inicios: [String.Index] = [texto.startIndex]
        let rango = NSRange(texto.startIndex..., in: texto)
        for match in corte.matches(in: texto, range: rango) {
            guard let corteRango = Range(match.range, in: texto) else { continue }
            let siguiente = corteRango.upperBound
            if siguiente < texto.endIndex, !inicios.contains(siguiente) {
                inicios.append(siguiente)
            }
        }
        var spans: [Range<String.Index>] = []
        for (i, start) in inicios.enumerated() {
            let end = i + 1 < inicios.count ? inicios[i + 1] : texto.endIndex
            if !texto[start..<end].trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
                spans.append(start..<end)
            }
        }
        return spans
    }

    private static func proteger(_ texto: String) -> (String, [String]) {
        guard let valla else { return (texto, []) }
        var held: [String] = []
        let ns = texto as NSString
        var piezas: [String] = []
        var cursor = 0
        valla.enumerateMatches(
            in: texto, range: NSRange(location: 0, length: ns.length)
        ) { match, _, _ in
            guard let match else { return }
            if match.range.location > cursor {
                piezas.append(ns.substring(
                    with: NSRange(location: cursor, length: match.range.location - cursor)
                ))
            }
            held.append(ns.substring(with: match.range))
            piezas.append("\u{0}H\(held.count - 1)\u{0}")
            cursor = match.range.location + match.range.length
        }
        if cursor < ns.length { piezas.append(ns.substring(from: cursor)) }
        return (piezas.joined(), held)
    }

    private static func restaurar(_ texto: String, held: [String]) -> String {
        var out = texto
        for (i, original) in held.enumerated() {
            out = out.replacingOccurrences(of: "\u{0}H\(i)\u{0}", with: original)
        }
        return out
    }

    private static func esPlaceholder(_ sentencia: String) -> Bool {
        sentencia.trimmingCharacters(in: .whitespacesAndNewlines)
            .range(of: #"^\u{0}H\d+\u{0}$"#, options: .regularExpression) != nil
    }
}

File: packages/core/edecan_core/speech_tags.py
"""Speech tags de Eleven v3: detección, densidad y relleno determinista.

El system prompt les pide a todos los modelos que etiqueten cada oración.
Muchos (sobre todo los rápidos) lo ignoran y dejan prosa plana. Esta capa
completa lo que falte **después** de generar, para que CUALQUIER LLM que
el dueño coloque suene humano al pulsar Escuchar.

No hay lista blanca de tags: [thoughtfully], [laughs], [applause],
[clears throat] o el efecto que invente el modelo son válidos. El chat
los oculta TODOS; la voz los usa tal cual. La rotación de abajo solo
rellena oraciones que el modelo dejó planas.
"""

from __future__ import annotations

import re

SPEECH_TAG_NAMES: tuple[str, ...] = (
    "warmly",
    "gently",
    "excited",
    "curious",
    "thoughtful",
    "serious",
    "playful",
    "calmly",
    "calm",
    "surprised",
    "empathetic",
    "pause",
    "whispering",
    "firmly",
    "amused",
    "reassuring",
    "hesitant",
)

# Cualquier [efecto] de una sola línea, corto. Conserva Markdown
# `[texto](url)` e `![alt](url)` — esos no son speech tags.
SPEECH_TAG_RE = re.compile(
    r"(?<!!)\[(?![^\]]{0,120}\]\()[^\]\n]{1,120}\]",
    re.I,
)

_FENCE_RE = re.compile(r"```.*?```", re.DOTALL)
_INLINE_CODE_RE = re.compile(r"`[^`]+`")
_MD_LINK_RE = re.compile(r"\[[^\]]+\]\([^)]+\)")
_MD_IMAGE_RE = re.compile(r"!\[[^\]]*\]\([^)]+\)")
_PLACEHOLDER_RE = re.compile(r"\x00H(\d+)\x00")
_BOUNDARY_RE = re.compile(r"[.!?;…]+[ \t]+|\n{2,}")

# Primera oración: tono, nunca [pause] suelto al arrancar.
_FIRST_DEFAULT = "warmly"
_ROTATION = ("pause", "thoughtful", "calm", "curious", "reassuring", "gently")

_CUE_TAGS: tuple[tuple[re.Pattern[str], str], ...] = (
    (
        re.compile(
            r"\b(lo siento|perdón|perdon|sorry|unfortunately|lástima|lastima)\b",
            re.I,
        ),
        "gently",
    ),
    (
        re.compile(
            r"\b(no (tengo|sé|se|estoy seguro)|not sure|i don.?t have|dato exacto|"
            r"informaci[oó]n exacta)\b",
            re.I,
        ),
        "hesitant",
    ),
    (
        re.compile(
            r"^(sin embargo|pero\b|aunque\b|however\b|but\b|aunque)",
            re.I,
        ),
        "thoughtful",
    ),
    (
        re.compile(
            r"\b(ten en cuenta|ojo\b|cuidado|importante|advertencia|"
            r"keep in mind|note that|warning)\b",
            re.I,
        ),
        "reassuring",
    ),
    (re.compile(r"[?¿]"), "curious"),
    (
        re.compile(
            r"[!¡]|\b(genial|increíble|excelente|wow|great|amazing)\b",
            re.I,
        ),
        "excited",
    ),
    (
        re.compile(
            r"\b(hola|hey|gracias|thank|claro|por supuesto|sure)\b",
            re.I,
        ),
        "warmly",
    ),
    (re.compile(r"[$€£%]|\d{2,}"), "thoughtful"),
)


def _proteger(text: str) -> tuple[str, list[str]]:
    held: list[str] = []

    def stash(match: re.Match[str]) -> str:
        held.append(match.group(0))
        return f"\x00H{len(held) - 1}\x00"

    value = _FENCE_RE.sub(stash, text)
    value = _MD_IMAGE_RE.sub(stash, value)
    value = _MD_LINK_RE.sub(stash, value)
    value = _INLINE_CODE_RE.sub(stash, value)
    return value, held


def _restaurar(text: str, held: list[str]) -> str:
    def unstash(match: re.Match[str]) -> str:
        return held[int(match.group(1))]

    return _PLACEHOLDER_RE.sub(unstash, text)


def _spans(text: str) -> list[tuple[int, int]]:
    if not text.strip():
        return []
    starts = [0]
    for match in _BOUNDARY_RE.finditer(text):
        nxt = match.end()
        if nxt < len(text) and nxt not in starts:
            starts.append(nxt)
    spans: list[tuple[int, int]] = []
    for i, start in enumerate(starts):
        end = starts[i + 1] if i + 1 < len(starts) else len(text)
        if text[start:end].strip():
            spans.append((start, end))
    return spans


def _elegir_tag(sentence: str, prev: str | None, *, first: bool) -> str:
    stripped = SPEECH_TAG_RE.sub("", sentence).strip()
    for pattern, tag in _CUE_TAGS:
        if pattern.search(stripped):
            if tag == prev:
                continue
            if first and tag == "pause":
                continue
            return tag
    if first:
        return _FIRST_DEFAULT if prev != _FIRST_DEFAULT else "calm"
    for tag in _ROTATION:
        if tag != prev:
            return tag
    return "thoughtful"


def ocultar_speech_tags(text: str) -> str:
    """Texto para pintar o copiar: sin tags ni efectos, con el Markdown intacto."""

    if not text:
        return text
    if not text.strip():
        return text
    protegido, held = _proteger(text)
    value = SPEECH_TAG_RE.sub("", protegido)
    value = re.sub(r"[ \t]{2,}", " ", value)
    value = re.sub(r"^[ \t]+", "", value)
    return _restaurar(value, held)


def enriquecer_speech_tags(text: str) -> str:
    """Devuelve el texto limpio, sin tags inventadas ni speech tags."""
    return ocultar_speech_tags(text)
